// @technote-space/github-action-pr-helper/dist/utils/variables.js
import { Utils, ContextHelper } from '@technote-space/github-action-helper';
import moment from 'moment';
import { getNewPatchVersion, getNewMinorVersion, getNewMajorVersion, getCurrentVersion, findPR, getDefaultBranch } from './command';
import { getActionDetail, getDefaultBranchUrl, getPrHeadRef, isActionPr, getContextBranch, isDefaultBranch, getActionContext, ensureGetPulls, getPullsArgsForDefaultBranch, getPrBranchPrefix, getPrBranchPrefixForDefaultBranch, isNotCreatePR, ParameterRequiredError, } from './misc';
const { getBranch } = Utils;
const { isPush } = ContextHelper;
export const getCommitMessage = (context) => getActionDetail('commitMessage', context);
export const getCommitName = (context) => getActionDetail('commitName', context, () => context.actionContext.actor);
export const getCommitEmail = (context) => getActionDetail('commitEmail', context, () => `${context.actionContext.actor}@users.noreply.github.com`);
const getVariable = (index, context) => getActionDetail('prVariables', context)[index];
export const getPrLink = (context) => context.actionContext.payload.pull_request ? `[${context.actionContext.payload.pull_request.title}](${context.actionContext.payload.pull_request.html_url})` : '';
const getDate = (index, context) => moment().format(getActionDetail('prDateFormats', context)[index]);
const contextVariables = async (isComment, octokit, context) => {
    const getContext = async (branch) => {
        if (isComment) {
            if (branch === await getDefaultBranch(octokit, context)) {
                return getActionContext(await getPullsArgsForDefaultBranch(octokit, context), octokit, context);
            }
            return getActionContext(await findPR(branch, octokit, context), octokit, context);
        }
        return context;
    };
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const getPrParamFunc = (extractor) => async () => {
        if (!context.actionContext.payload.pull_request) {
            throw new Error('Invalid context.');
        }
        return extractor(await ensureGetPulls((await getContext(getContextBranch(context))).actionContext.payload.pull_request, octokit, context));
    };
    return [
        { key: 'PR_NUMBER', replace: getPrParamFunc(pr => pr.number) },
        { key: 'PR_NUMBER_REF', replace: getPrParamFunc(async (pr) => pr.number ? `#${pr.number}` : await getDefaultBranchUrl(octokit, context)) },
        { key: 'PR_ID', replace: getPrParamFunc(pr => pr.id) },
        { key: 'PR_HEAD_REF', replace: getPrParamFunc(pr => pr.head.ref) },
        { key: 'PR_BASE_REF', replace: getPrParamFunc(pr => pr.base.ref) },
        { key: 'PR_TITLE', replace: getPrParamFunc(pr => pr.title) },
        { key: 'PR_URL', replace: getPrParamFunc(pr => pr.html_url) },
        { key: 'PR_MERGE_REF', replace: getPrParamFunc(async (pr) => pr.number ? `${pr.head.ref} -> ${pr.base.ref}` : await getDefaultBranch(octokit, context)) },
        { key: 'PR_LINK', replace: async () => getPrLink(context) },
        { key: 'CURRENT_VERSION', replace: async () => getCurrentVersion(octokit, context) },
        { key: 'PATCH_VERSION', replace: async () => getNewPatchVersion(octokit, context) },
        { key: 'MINOR_VERSION', replace: async () => getNewMinorVersion(octokit, context) },
        { key: 'MAJOR_VERSION', replace: async () => getNewMajorVersion(octokit, context) },
        // eslint-disable-next-line no-magic-numbers
    ].concat([...Array(context.actionDetail.prVariables?.length ?? 0).keys()].map(index => ({
        // eslint-disable-next-line no-magic-numbers
        key: `VARIABLE${index + 1}`, replace: async () => getVariable(index, context),
        // eslint-disable-next-line no-magic-numbers
    }))).concat([...Array(context.actionDetail.prDateFormats?.length ?? 0).keys()].map(index => ({
        // eslint-disable-next-line no-magic-numbers
        key: `DATE${index + 1}`, replace: async () => getDate(index, context),
    })));
};
const replaceContextVariables = async (string, octokit, context) => Utils.replaceVariables(string, await contextVariables(false, octokit, context));
export const getPrBranchName = async (octokit, context, isDuplicateCheck = false) => {
    if (isPush(context.actionContext)) {
        return getBranch(context.actionContext);
    }
    if (isActionPr(context) || isNotCreatePR(context)) {
        return getPrHeadRef(context);
    }
    let prefix;
    let branch;
    if (await isDefaultBranch(octokit, context)) {
        prefix = getPrBranchPrefixForDefaultBranch(context);
    }
    else {
        prefix = getPrBranchPrefix(context);
    }
    try {
        if (await isDefaultBranch(octokit, context)) {
            branch = getActionDetail('prBranchNameForDefaultBranch', context, () => getActionDetail('prBranchName', context));
        }
        else {
            branch = getActionDetail('prBranchName', context);
        }
    }
    catch (error) {
        if (isDuplicateCheck && (error instanceof ParameterRequiredError)) {
            return `${context.actionContext.runNumber}`;
        }
        throw error;
    }
    return prefix + await replaceContextVariables(branch, octokit, context);
};
export const getPrTitle = async (octokit, context) => await replaceContextVariables((await isDefaultBranch(octokit, context) ?
    getActionDetail('prTitleForDefaultBranch', context, () => getActionDetail('prTitle', context)) :
    getActionDetail('prTitle', context)).trim(), octokit, context);
const prBodyVariables = async (isComment, files, output, octokit, context) => {
    const toCode = (string) => string.length ? ['', '```Shell', string, '```', ''].join('\n') : '';
    return [
        {
            key: 'COMMANDS',
            replace: async () => output.length ? toCode(output.map(item => `$ ${item.command}`).join('\n')) : '',
        },
        {
            key: 'COMMANDS_STDOUT',
            replace: async () => output.length ? '<details>\n' + output.map(item => [
                `<summary><em>${item.command}</em></summary>`,
                toCode(item.stdout.join('\n')),
            ].join('\n')).join('\n</details>\n<details>\n') + '\n</details>' : '',
        },
        {
            key: 'COMMANDS_OUTPUT',
            replace: async () => output.length ? '<details>\n' + output.map(item => [
                `<summary><em>${item.command}</em></summary>`,
                toCode(item.stdout.join('\n')),
                item.stderr.length ? '### stderr:' : '',
                toCode(item.stderr.join('\n')),
            ].join('\n')).join('\n</details>\n<details>\n') + '\n</details>' : '',
        },
        {
            key: 'COMMANDS_STDOUT_OPENED',
            replace: async () => output.length ? '<details open>\n' + output.map(item => [
                `<summary><em>${item.command}</em></summary>`,
                toCode(item.stdout.join('\n')),
            ].join('\n')).join('\n</details>\n<details open>\n') + '\n</details>' : '',
        },
        {
            key: 'COMMANDS_STDERR',
            replace: async () => output.length ? '<details>\n' + output.map(item => [
                `<summary><em>${item.command}</em></summary>`,
                toCode(item.stderr.join('\n')),
            ].join('\n')).join('\n</details>\n<details>\n') + '\n</details>' : '',
        },
        {
            key: 'COMMANDS_STDERR_OPENED',
            replace: async () => output.length ? '<details open>\n' + output.map(item => [
                `<summary><em>${item.command}</em></summary>`,
                toCode(item.stderr.join('\n')),
            ].join('\n')).join('\n</details>\n<details open>\n') + '\n</details>' : '',
        },
        {
            key: 'FILES',
            replace: async () => files.map(file => `- ${file}`).join('\n'),
        },
        {
            key: 'FILES_SUMMARY',
            // eslint-disable-next-line no-magic-numbers
            replace: async () => 'Changed ' + (files.length > 1 ? `${files.length} files` : 'file'),
        },
        {
            key: 'ACTION_NAME',
            replace: async () => context.actionDetail.actionName,
        },
        {
            key: 'ACTION_OWNER',
            replace: async () => context.actionDetail.actionOwner,
        },
        {
            key: 'ACTION_REPO',
            replace: async () => context.actionDetail.actionRepo,
        },
        {
            key: 'ACTION_URL',
            replace: async () => `https://github.com/${context.actionDetail.actionOwner}/${context.actionDetail.actionRepo}`,
        },
        {
            key: 'ACTION_MARKETPLACE_URL',
            replace: async () => `https://github.com/marketplace/actions/${context.actionDetail.actionRepo}`,
        },
    ].concat(await contextVariables(isComment, octokit, context));
};
const replacePrBodyVariables = async (isComment, prBody, files, output, octokit, context) => Utils.replaceVariables(prBody, await prBodyVariables(isComment, files, output, octokit, context));
export const getPrBody = async (isComment, files, output, octokit, context) => replacePrBodyVariables(isComment, (isComment ?
    getActionDetail('prBodyForComment', context, () => getActionDetail('prBody', context)) :
    (await isDefaultBranch(octokit, context) ?
        getActionDetail('prBodyForDefaultBranch', context, () => getActionDetail('prBody', context)) :
        getActionDetail('prBody', context))).trim().split(/\r?\n/).map(line => line.replace(/^[\s\t]+/, '')).join('\n'), files, output, octokit, context);
