// @technote-space/doctoc/build/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SEPARATOR = exports.DEFAULT_ITEM_TEMPLATE = exports.DEFAULT_CUSTOM_TEMPLATE = exports.IGNORED_DIRS = exports.MARKDOWN_EXTENSIONS = exports.DEFAULT_TITLE = exports.CHECK_CLOSING_COMMENT = exports.CHECK_OPENING_COMMENT = exports.CLOSING_COMMENT = exports.OPENING_COMMENT = void 0;
exports.OPENING_COMMENT = '<!-- START doctoc generated TOC please keep comment here to allow auto update -->\n' +
    '<!-- DON\'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->';
exports.CLOSING_COMMENT = '<!-- END doctoc generated TOC please keep comment here to allow auto update -->';
exports.CHECK_OPENING_COMMENT = '<!-- START doctoc ';
exports.CHECK_CLOSING_COMMENT = '<!-- END doctoc ';
exports.DEFAULT_TITLE = '**Table of Contents**  *generated with [DocToc](https://github.com/technote-space/doctoc)*';
exports.MARKDOWN_EXTENSIONS = ['.md', '.markdown'];
exports.IGNORED_DIRS = ['.', '..', '.git', 'node_modules'];
exports.DEFAULT_CUSTOM_TEMPLATE = '<p align="center">${ITEMS}</p>';
exports.DEFAULT_ITEM_TEMPLATE = '<a href="${LINK}">${TEXT}</a>';
exports.DEFAULT_SEPARATOR = '<span>|</span>';
